### PR TITLE
Sync - project setup, sync module and sync test module

### PR DESCRIPTION
**Pull request by edualonso**
- Added `library-sync` module under `packages` and configured it to depend on `library-base` (former `library`).
- The `test` module now hosts a `base` module which contains all previously existing tests that target local db use cases and a `sync` module for sync-related tests. For now `sync` only contains a sample test.

TODO:
- [x] CI: run sync tests
- [ ] Dokka: publish documentation only for sync https://github.com/realm/realm-kotlin/issues/394
- [ ] Release: analyse release process to include changes for sync https://github.com/realm/realm-kotlin/issues/421

### CODE REVIEW

File: packages/settings.gradle.kts [hunk 1]
```diff
@@ -19,6 +19,7 @@ include("gradle-plugin")
 include("plugin-compiler")
 include("plugin-compiler-shaded")
 include("library")
+include("sync")
```
Comment (cmelchior) [inline packages/settings.gradle.kts:22]: Maybe `sync-library` and `base-library`? Just to make it clear what they actually are?

File: packages/sync/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+            }
+            jdkVersion.set(8)
+        }
+        val commonMain by getting {
+            sourceRoot("../runtime-api/src/commonMain/kotlin")
+        }
+    }
+}
+
+tasks.register("uploadDokka") {
```
Comment (cmelchior) [inline packages/sync/build.gradle.kts:248]: I suspect we want to disable this in the base-library and only upload these from here. But maybe that is going to be done in another PR. If so, we should probably create an issue so we remember it.

File: packages/sync/proguard-rules-consumer-common.pro [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+-keep class * implements io.realm.RealmObject { *; }
+#-keep class **.$* implements io.realm.RealmObject { *; }
+
+# Preserve all native method names and the names of their classes.
+-keepclasseswithmembernames,includedescriptorclasses class * {
+    native <methods>;
+}
+
+# Notification callback
+-keep class io.realm.interop.NotificationCallback {
```
Comment (cmelchior) [inline packages/sync/proguard-rules-consumer-common.pro:12]: Do we need these here in the sync variant? I assumed they would be inherited?

File: packages/sync/src/androidMain/kotlin/io/realm/sync/transport/NetworkTransport.kt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+package io.realm.sync.transport
```
Comment (cmelchior) [inline packages/sync/src/androidMain/kotlin/io/realm/sync/transport/NetworkTransport.kt:1]: In Java, this package is called `io.realm.mongodb` with a `io.realm.mongodb.sync` subpackage. IMO we should do the same here.

File: test/settings.gradle.kts [hunk 5]
```diff
@@ -9,24 +9,26 @@
...
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
- * or implied.
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
 
 includeBuild("../packages")
 
+include("local")
```
Comment (cmelchior) [inline test/settings.gradle.kts:19]: What was your thinking here? To have a `local` and `sync` package with completely separate tests? Or let a `sync` module depend on the `local` module?

In Java we run the local tests for both sync and local builds as they are using two different `.so` files. It does feel a bit extreme, but I suspect we need to do something similar here to be sure we actually test what we ship?

File: packages/sync/proguard-rules-consumer-common.pro (hunk 3 above)
Comment (edualonso) [reply packages/sync/proguard-rules-consumer-common.pro:12]: True, forgot about proguard cleanup.

File: packages/sync/build.gradle.kts (hunk 2 above)
Comment (edualonso) [reply packages/sync/build.gradle.kts:248]: Sure, just created https://github.com/realm/realm-kotlin/issues/394

File: packages/sync/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,303 @@
+import io.realm.getPropertyValue
```
Comment (rorbech) [inline packages/sync/build.gradle.kts:1]: You should redo this from the kmm-configuration cleaned-up one from `library`.

File: packages/sync/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+//        exec {
+//            commandLine("xcrun", "simctl", "spawn", device, binary.absolutePath)
+//        }
+//    }
+// }
+
+realmPublish {
+    pom {
+        name = "Sync"
+        description = "Sync code for Realm Kotlin. This artifact is not " +
```
Comment (rorbech) [inline packages/sync/build.gradle.kts:222]: I guess this is not true at the moment. We have a `TODO AUTO-SETUP` for these things that needs to be aligned when we inject the dependencies.

File: packages/sync/gradlew.bat [hunk 8]
```diff
@@ -0,0 +1,100 @@
+@rem
```
Comment (rorbech) [inline packages/sync/gradlew.bat:1]: This file is needed?

File: packages/sync/gradle.properties [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#
```
Comment (rorbech) [inline packages/sync/gradle.properties:1]: I think we should just have these settings in the `packages` specific file?

File: packages/sync/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+            "supposed to be consumed directly, but through " +
+            "'io.realm.kotlin:gradle-plugin:${Realm.version}' instead."
+    }
+}
+
+tasks.dokkaHtml.configure {
```
Comment (rorbech) [inline packages/sync/build.gradle.kts:228]: I guess there is more to Dokka-generation as we would need one artifact containing both `library-base` and `library-sync`. Don't really know how to do this when it is in different modules. So we should probably have a complete separate issue to do the dokka-generation.

File: packages/settings.gradle.kts (hunk 1 above)
Comment (rorbech) [reply packages/settings.gradle.kts:22]: Maybe `library-base` and `library-sync` just to keep the source trees as neighbours in the IDE?

File: packages/sync/gradlew.bat (hunk 8 above)
Comment (rorbech) [inline packages/sync/gradlew.bat:1]: We don't need this!?

File: packages/sync/gradle.properties [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+kotlin.code.style=official
```
Comment (rorbech) [inline packages/sync/gradle.properties:18]: Couldn't we just have these settings in the `packages`-level properties file?

File: packages/sync/src/androidMain/AndroidManifest.xml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    package="io.realm.sync">
+
+    <application>
+        <provider
```
Comment (rorbech) [inline packages/sync/src/androidMain/AndroidManifest.xml:23]: We don't need to put a startup provider here unless we really need on. Manifest-merging will include the one from `library-base`.

File: packages/sync/src/jvmMain/kotlin/io/realm/sync/transport/NetworkTransport.kt [hunk 13]
```diff
@@ -0,0 +1,2 @@
+package io.realm.sync.transport
```
Comment (rorbech) [inline packages/sync/src/jvmMain/kotlin/io/realm/sync/transport/NetworkTransport.kt:1]: Can't we have a common one for Android and JVM in `src/jvm/kotlin` instead? See the updated kmm-setup for `library` in master. 

Maybe there are some `ktor` dependencies that aren't resolvable if we do it that way, but then it might work by just adding `src/jvm/kotlin` to both `androidMain` and `jvmMain` source sets.

File: test/build.gradle.kts [hunk 14]
```diff
@@ -9,185 +9,27 @@
...
-import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTargetWithSimulatorTests
-
 plugins {
-    id("org.jetbrains.kotlin.multiplatform")
-    id("com.android.library")
-    id("io.realm.kotlin")
+    id("com.android.library") apply false
     id("realm-lint")
+    `java-gradle-plugin`
+    id("realm-publisher")
```
Comment (rorbech) [inline test/build.gradle.kts:21]: Why do we need the publisher in test?

File: test/build.gradle.kts [hunk 15]
```diff
@@ -9,185 +9,27 @@
...
-    }
-}
 
-// Needs running emulator
-tasks.named("iosTest") {
-    val device: String = project.findProperty("iosDevice")?.toString() ?: "iPhone 11 Pro Max"
-    dependsOn(kotlin.targets.getByName<KotlinNativeTargetWithSimulatorTests>("ios").binaries.getTest("DEBUG").linkTaskName)
-    group = JavaBasePlugin.VERIFICATION_GROUP
-    description = "Runs tests for target 'ios' on an iOS simulator"
+    version = Realm.version
```
Comment (rorbech) [inline test/build.gradle.kts:29]: Why do we need these?

File: test/local/gradle.properties [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+org.gradle.jvmargs=-Xms512m -Xmx2048m
```
Comment (rorbech) [inline test/local/gradle.properties:18]: Put these in `test/gradle.properties` instead of having to replicate them.

File: test/local/gradlew [hunk 17]
```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env sh
```
Comment (rorbech) [inline test/local/gradlew:1]: Not needed.

File: test/settings.gradle.kts (hunk 5 above)
Comment (rorbech) [reply test/settings.gradle.kts:19]: I advices that we should have a separate module for `base` and `sync`. Didn't know that we actually did `base` test on `sync` too. If we initially ship the same `.so`-file it isn't blocking, but it would be a good idea to evaluate if we need to rethink the split.

File: test/settings.gradle.kts [hunk 18]
```diff
@@ -9,24 +9,26 @@
...
+
 pluginManagement {
+    plugins {
+    }
```
Comment (rorbech) [inline test/settings.gradle.kts:23]:
```suggestion
```

**Review by rorbech — CHANGES_REQUESTED** [review]
Added some comments here and there. Most importantly you should use the cleanup multiplatform setup from master as a base for this.

File: packages/settings.gradle.kts (hunk 1 above)
Comment (cmelchior) [reply packages/settings.gradle.kts:22]: Yeah, that would be better 👍

Comment (cmelchior) [reply packages/settings.gradle.kts:22]: Yeah, that would be better 👍

File: packages/sync/src/jvmMain/kotlin/io/realm/sync/transport/NetworkTransport.kt (hunk 13 above)
Comment (edualonso) [reply packages/sync/src/jvmMain/kotlin/io/realm/sync/transport/NetworkTransport.kt:1]: I haven't added anything related to Ktor in this PR yet, it will come in the next one so I'll have a look at that then.

File: test/build.gradle.kts (hunk 14 above)
Comment (edualonso) [reply test/build.gradle.kts:21]: Wrong copy-paste.

File: test/build.gradle.kts (hunk 15 above)
Comment (edualonso) [reply test/build.gradle.kts:29]: Wrong copy-paste.

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/NetworkTransport.kt [hunk 19]
```diff
@@ -0,0 +1,3 @@
+package io.realm.mongodb.sync
```
Comment (rorbech) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/NetworkTransport.kt:1]: Internal? Or is this actually needed by the user for setup?

File: test/local/build.gradle.kts [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+    google()
+    jcenter()
+    mavenCentral()
+    mavenLocal()
+}
+
+// Common Kotlin configuration
+kotlin {
+    sourceSets {
+        getByName("commonMain") {
```
Comment (rorbech) [inline test/local/build.gradle.kts:37]: Hmm. Apparently I missed cleaning up the configuration for the tests. We should probably just leave that for another PR as it will be easier to track the renaming if we not rework the configuration completely.

**Review by rorbech — COMMENTED** [review]
Looks better, but did you forget to add `test/sync`?

Also seems weird with `test/local` when the library is called `library-base`. So maybe `test/base` and `test/sync`?

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/NetworkTransport.kt (hunk 19 above)
Comment (edualonso) [reply packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/NetworkTransport.kt:1]: I believe it should be internal, yes.

**Comment by edualonso** [comment]
`test/sync` will be added in the next PR.

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/NetworkTransport.kt (hunk 19 above)
Comment (edualonso) [reply packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/sync/NetworkTransport.kt:1]: Now that I think of it, I don't know if it can be internal since it will be the C-API that will be calling the `sendRequest` function. I guess we will have to try whether it can be done 🤷

**Comment by edualonso** [comment]
I agree with your module naming comment but implementing that change in this PR would pretty much flood my upcoming PR with conflicts so I will do it in the next one.

File: test/sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncDispatcher.kt [hunk 21]
```diff
@@ -0,0 +1,5 @@
+package io.realm.mongodb.sync
+
+import kotlinx.coroutines.CoroutineDispatcher
+
+expect fun getDispatcher(): CoroutineDispatcher
```
Comment (rorbech) [inline test/sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncDispatcher.kt:5]: You could just use `io.realm.internal.platform.singleThreadDispatcher` from the library?

**Review by rorbech — COMMENTED** [review]
Overall LGTM - Some comments:
- Why do we need platform differentiated dispatcher specifically for sync?
- Have you verified that things are still in place for `base`? 
  - All tests are actually run
  - Dokka for `base` is still generated/published correctly

If things are still in place for `base` I think we can just track the remainders in new issues.

File: test/sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncDispatcher.kt (hunk 21 above)
Comment (edualonso) [reply test/sync/src/commonMain/kotlin/io/realm/mongodb/sync/SyncDispatcher.kt:5]: Yes, that's what I'm doing in the next PR.

**Comment by edualonso** [comment]
There will be some cleanup related to the comments in the PR we will put up next.
Tests for base are run, just checked the build log for build no. 14 and everything looks normal.
Dokka seems to be working and generates the artifact correctly, but the docs format is not correct. @rorbech says it is likely this issue has been introduced after the build script cleanup task.
Apart from that nothing that I'm aware of.

**Review by rorbech — APPROVED** [review]

**Comment by rorbech** [comment]
> Dokka seems to be working and generates the artifact correctly, but the docs format is not correct. @rorbech says it is likely this issue has been introduced after the build script cleanup task.

Created #438 to fix this separately as it doesn't seem related to this PR.
